episodes/12-data-cube.md

```diff
@@ -0,0 +1,195 @@
+    red          (time, y, x) uint16 418MB dask.array<chunksize=(1, 2048, 2048), meta=np.ndarray>
+    green        (time, y, x) uint16 418MB dask.array<chunksize=(1, 2048, 2048), meta=np.ndarray>
+    blue         (time, y, x) uint16 418MB dask.array<chunksize=(1, 2048, 2048), meta=np.ndarray>
+    visual       (time, y, x) float32 836MB dask.array<chunksize=(1, 2048, 2048), meta=np.ndarray>
+    nir          (time, y, x) uint16 418MB dask.array<chunksize=(1, 2048, 2048), meta=np.ndarray>
+    swir22       (time, y, x) uint16 418MB dask.array<chunksize=(1, 2048, 2048), meta=np.ndarray>
+    ...           ...
+    scl          (time, y, x) uint8 209MB dask.array<chunksize=(1, 2048, 2048), meta=np.ndarray>
+    aot          (time, y, x) uint16 418MB dask.array<chunksize=(1, 2048, 2048), meta=np.ndarray>
+    coastal      (time, y, x) uint16 418MB dask.array<chunksize=(1, 2048, 2048), meta=np.ndarray>
+    nir09        (time, y, x) uint16 418MB dask.array<chunksize=(1, 2048, 2048), meta=np.ndarray>
+    cloud        (time, y, x) uint8 209MB dask.array<chunksize=(1, 2048, 2048), meta=np.ndarray>
+    snow         (time, y, x) uint8 209MB dask.array<chunksize=(1, 2048, 2048), meta=np.ndarray>
+```
+
+## Working with the data cube
+
+Like we did in the previous episode, let us calculate the NDVI for our study area. To do so we need to focus on the variables: the red band (`red`), the near infrared band (`nir`) and the scene classification map (`scl`). We will use the former two to calculated the NDVI for the AoI. The latter, we use as [a classification mask](https://sentinels.copernicus.eu/web/sentinel/technical-guides/sentinel-2-msi/level-2a/algorithm-overview) provided together with Sentinel-2 L2A products. In this mask, each pixel is classified according to a set of labels (see Figure 3 in [classification mask documentation](https://sentinels.copernicus.eu/web/sentinel/technical-guides/sentinel-2-msi/level-2a/algorithm-overview) ). 
+
+First we define the bands that we are interested in:
+
+```python
+red = ds['red']
+nir = ds['nir']
+scl = ds['scl']
+```
+
+Next we will use the mask to drop pixels that are labeled as clouds and water. For this we use the classification map to mask out pixels recognized by the [Sentinel-2 processing algorithm](https://custom-scripts.sentinel-hub.com/custom-scripts/sentinel-2/scene-classification/) as cloud or water:
+
+```python
+# generate mask ("True" for pixel being cloud or water)
+mask = scl.isin([
+    3,  # CLOUD_SHADOWS
+    6,  # WATER
+    8,  # CLOUD_MEDIUM_PROBABILITY
+    9,  # CLOUD_HIGH_PROBABILITY
+    10  # THIN_CIRRUS
+])
+red_masked = red.where(~mask)
+nir_masked = nir.where(~mask)
+```
+
+Then, we calculate the NDVI:
+
+```python
+ndvi = (nir_masked - red_masked) / (nir_masked + red_masked)
+```
+
+We can visualize the calculated NDVI for the AoI at two given dates (before and after the wildfires) by selecting the date:
+
+```python
+ndvi_before = ndvi.sel(time="2023-07-13")
```

Comment:
this cell that triggers the computation presents a warning

```
/Users/ryanavery/test-dask-on-ray/.venv/lib/python3.11/site-packages/rasterio/warp.py:387: NotGeoreferencedWarning: Dataset has no geotransform, gcps, or rpcs. The identity matrix will be returned.
  dest = _reproject(
```

we should diagnose why georeferencing is not preserved and try to preserve it in this example so that geospatial metadata can be roundtripped when saving and loading the ndvi raster

Reply:
though I'm testing this in an env called "dask-on-ray" this is just with vanilla dask and xarray. here are my versions

```
    "odc-stac>=0.3.10",
    "pystac>=1.11.0",
    "pystac-client>=0.8.3",
    "ray>=2.37.0",
    "rioxarray>=0.17.0",
    "torch>=2.4.1",
    "xarray>=2024.9.0",
    "leafmap>=0.38.3",
    "geopandas>=1.0.1",
```